core/ir/synthetic/encode.c

```diff
@@ -0,0 +1,129 @@
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* encode.c -- an encoder for DR synthetic IR */
+
+#include "encode.h"
+
+#include "../globals.h"
+#include "encode_api.h"
+#include "encoding_common.h"
+#include "instr_api.h"
+#include "opnd_api.h"
+
+/* Encodes DR instruction representation \p instr into raw bytes \p encoded_instr.
+ * Returns next instruction's pc.
+ * The encoding scheme followed is described in core/ir/synthetic/encoding_common.h.
+ */
+byte *
+encode_to_synth(dcontext_t *dcontext, instr_t *instr, byte *encoded_instr)
+{
+    /* Use a local uint variable for easier setting of category, eflags, #src, and #dst
+     * values.
+     */
+    uint encoding_header = 0;
+
+    /* Encode number of register destination operands (i.e., written registers).
+     */
+    uint num_dsts = (uint)instr_num_dsts(instr);
+    encoding_header |= num_dsts;
+
+    /* Encode number of register source operands (i.e., read registers).
+     */
+    uint num_srcs = (uint)instr_num_srcs(instr);
+    encoding_header |= (num_srcs << SRC_OPND_SHIFT);
+
+    /* Encode arithmetic flags.
+     */
+    uint eflags_instr = instr_get_arith_flags(instr, DR_QUERY_DEFAULT);
+    uint eflags = 0;
+    if (TESTANY(EFLAGS_WRITE_ARITH, eflags_instr))
+        eflags |= SYNTHETIC_INSTR_WRITES_ARITH;
+    if (TESTANY(EFLAGS_READ_ARITH, eflags_instr))
+        eflags |= SYNTHETIC_INSTR_READS_ARITH;
+    encoding_header |= (eflags << FLAGS_SHIFT);
+
+    /* Encode category as synthetic opcode.
+     */
+    uint category = instr_get_category(instr);
+    encoding_header |= (category << CATEGORY_SHIFT);
+
+    /* Copy header encoding back into encoded_instr output.
+     */
+    *((uint *)&encoded_instr[0]) = encoding_header;
+
+    /* Encode register destination operands, if present.
+     */
+    opnd_size_t max_opnd_size = OPSZ_NA;
+    for (uint dst_index = 0; dst_index < num_dsts; ++dst_index) {
+        opnd_t dst_opnd = instr_get_dst(instr, dst_index);
+        max_opnd_size = opnd_get_size(dst_opnd);
```

Comment:
Hmm, ok so this is where the max opnd size including the memops breaks down: b/c that max size is not stored in the IR.

But if you store it in the reg opnds themselves, they would become inconsistent and not match the reg id.

Discuss offline?

Reply:
Yes, let's discuss offline.
Just a note: register ids in the REGDEPS ISA are really just numbers (identifiers), so if you're referring to consistency between the operand size and the size you'd get when calling `reg_get_size()` on REGDEPS registers, then you're right, they wouldn't match, and I'd argue we don't want them to match, because REGDEPS registers are not "real register" (as REGDEPS is not a real ISA), it just so happens that their ids are within the DR_REG_ range.
It would be nice to have `reg_get_size()` fail for REGDEPS registers though.
Analyses should use `opnd_get_size()`, in which case you'd get the correct answer.

Reply:
As discussed offline, we now take the max source operand size (regardless of operands being registers or anything else).
Done.